homeassistant/components/zwave_js/discovery_data_template.py

```diff
@@ -189,43 +324,67 @@ def current_temperature_value(resolved_data: dict[str, Any]) -> ZwaveValue | Non
 class NumericSensorDataTemplate(BaseDiscoverySchemaDataTemplate):
     """Data template class for Z-Wave Sensor entities."""
 
-    def resolve_data(self, value: ZwaveValue) -> str | None:
+    @staticmethod
+    def find_key_from_matching_set(
+        enum_value: MultilevelSensorType | MultilevelSensorScaleType | MeterScaleType,
+        set_map: dict[
+            str, set[MultilevelSensorType | MultilevelSensorScaleType | MeterScaleType]
+        ],
+    ) -> str | None:
+        """Find a key in a set map that matches a given enum value."""
+        for key, value_set in set_map.items():
+            for value_in_set in value_set:
+                # Since these are IntEnums and the different classes reuse the same
+                # values, we need to match the class as well
+                if (
+                    value_in_set.__class__ == enum_value.__class__
+                    and value_in_set == enum_value
+                ):
+                    return key
+        return None
+
+    def resolve_data(self, value: ZwaveValue) -> tuple[str | None, str | None]:
         """Resolve helper class data for a discovered value."""
 
         if value.command_class == CommandClass.BATTERY:
-            return ENTITY_DESC_KEY_BATTERY
+            return (ENTITY_DESC_KEY_BATTERY, PERCENTAGE)
 
         if value.command_class == CommandClass.METER:
             scale_type = get_meter_scale_type(value)
+            unit = self.find_key_from_matching_set(scale_type, METER_UNIT_MAP)
             # We do this because even though these are energy scales, they don't meet
             # the unit requirements for the energy device class.
             if scale_type in (
                 ElectricScale.PULSE_COUNT,
                 ElectricScale.KILOVOLT_AMPERE_HOUR,
                 ElectricScale.KILOVOLT_AMPERE_REACTIVE_HOUR,
             ):
-                return ENTITY_DESC_KEY_TOTAL_INCREASING
+                return (ENTITY_DESC_KEY_TOTAL_INCREASING, unit)
             # We do this because even though these are power scales, they don't meet
             # the unit requirements for the power device class.
             if scale_type == ElectricScale.KILOVOLT_AMPERE_REACTIVE:
-                return ENTITY_DESC_KEY_MEASUREMENT
+                return (ENTITY_DESC_KEY_MEASUREMENT, unit)
 
-            for key, scale_type_set in METER_DEVICE_CLASS_MAP.items():
-                if scale_type in scale_type_set:
-                    return key
+            return (
+                self.find_key_from_matching_set(scale_type, METER_DEVICE_CLASS_MAP),
```

Comment:
Maybe we should return a dataclass or named tuple with the items as two attributes to make it more clear what the return values are?